### PR TITLE
Fix/saving cues on timeline

### PR DESCRIPTION
**Pull request by rubenwiersma**
Cues are now saved and added to the database when they are edited on the timeline. Also, some visual changes.

### CODE REVIEW

File: src/main/webapp/app/scripting/timeline.directive.js [hunk 1]
```diff
@@ -128,38 +194,80 @@
...
+                    backdrop: 'static',
+                    size: 'md',
+                    resolve: {
+                        entity: item.cue
+                    }
+                }).result.then(function(result) {
+                    console.log('result is: ', result);
+                    callback(item);
+                }, function() {
+                    console.log('cancelled');
```
Comment (hungnguyen94) [inline src/main/webapp/app/scripting/timeline.directive.js]: I forgot to put `callback(null);` here. It's needed to cancel the action.

**Comment by hungnguyen94** [comment]
`index.html` :angry:

**Comment by hungnguyen94** [comment]
This also fixes the issue where clicking on the map would not uncollapse the table until you moved the entity. Now clicking on a map entity will immediately uncollapse the right entry in the table.

**Comment by hungnguyen94** [comment]
The update button in the Map Editor can be removed now.

**Comment by rubenwiersma** [comment]
#89

File: src/main/webapp/app/map/fabric-map.directive.js [hunk 2]
```diff
@@ -7,23 +7,41 @@
...
         return directive;
 
+        /**
+         * Links the controller with the directive.
+         * @param scope the scope that the directive is called in
+         * @param element the element that the directive is called in
+         * @param attrs the attributes of the element that the directive is called in
+         */
         function link(scope, element, attrs) {
             console.log('fabric map directive called');
```
Comment (cadw) [inline src/main/webapp/app/map/fabric-map.directive.js]: I don't think we will need this console.log() anymore.

File: src/main/webapp/app/map/fabric-map.directive.js [hunk 3]
```diff
@@ -42,11 +60,13 @@
 
             /////////////////////////
 
-
+            /**
+             * Initializes the map with a canvas and loads the cameras and players.
+             */
             function init() {
                 console.log('init directive');
```
Comment (cadw) [inline src/main/webapp/app/map/fabric-map.directive.js]: Can be removed.

File: src/main/webapp/app/map/fabric-map.directive.js [hunk 4]
```diff
@@ -59,7 +79,6 @@
                     resize();
                 });
                 scope.canvas = canvas;
-                // drawGrid(grid);
```
Comment (cadw) [inline src/main/webapp/app/map/fabric-map.directive.js]: Can be removed, or if you want to keep it, you can add a global boolean for enabling.disabling the grid.

File: src/main/webapp/app/map/fabric-map.directive.js [hunk 5]
```diff
@@ -80,15 +99,25 @@
                 scope.canvas.renderAll();
                 scope.canvas.calcOffset();
                 console.log('parent is ', parent);
```
Comment (cadw) [inline src/main/webapp/app/map/fabric-map.directive.js]: Console.log() can be removed.

File: src/main/webapp/app/map/fabric-map.directive.js [hunk 6]
```diff
@@ -98,18 +127,33 @@
                 scope.canvas.renderAll();
             }
 
+            /**
+             * Sets the properties of the draw object.
+             * @param drawable the object that is drawn
+             * @returns {*}
+             */
             function setDrawableProperties(drawable) {
                 drawable.width = 15;
```
Comment (cadw) [inline src/main/webapp/app/map/fabric-map.directive.js]: Do we still need the width if the grid is removed?

File: src/main/webapp/app/map/fabric-map.directive.js [hunk 7]
```diff
@@ -98,18 +127,33 @@
...
                 return drawable;
             }
 
+            /**
+             * Transforms a camera and sets its attributes.
+             * @param camera the camera to transform
+             * @returns {*}
+             */
             function transformCamera(camera) {
                 var cam = new fabric.Path('m96.6,26.8h-86.1c-2.2,0-4.1,1.8-4.1,4.1v67.2c0,2.2 1.8,4.1 4.1,4.1h86.1c2.2,0 4.1-1.8 4.1-4.1v-19.4l14.9,14.9c0.8,0.8 1.8,1.2 2.9,1.2 0.5,0 1.1-0.1 1.6-0.3 1.5-0.6 2.5-2.1 2.5-3.8v-52.5c0-1.6-1-3.1-2.5-3.8-1.5-0.6-3.3-0.3-4.4,0.9l-14.9,14.9v-19.3c-0.1-2.3-1.9-4.1-4.2-4.1zm-4.1,33.3v8.8 25.2h-78v-59.2h78v25.2zm21.9-12v32.9l-13.7-13.7v-5.4l13.7-13.8z');
```
Comment (cadw) [inline src/main/webapp/app/map/fabric-map.directive.js]: Looks hardcoded :) Is there another way to do this?

File: src/main/webapp/app/map/fabric-map.directive.js [hunk 8]
```diff
@@ -187,18 +251,28 @@
                 }
             }
 
+            /**
+             * Displays the information of an object when it is selected.
+             * @param options
+             */
             function onSelect (options) {
                 console.log('Selected object: ', options);
```
Comment (cadw) [inline src/main/webapp/app/map/fabric-map.directive.js]: Can be removed

File: src/main/webapp/app/map/fabric-map.directive.js [hunk 9]
```diff
@@ -175,6 +235,10 @@
                 return position;
             }
 
+            /**
+             * Updates an entity (camera or player) when it is moved on the map.
+             * @param options the options that are changed
+             */
             function updateEntity(options) {
                 console.log('Modified object: ', options);
```
Comment (cadw) [inline src/main/webapp/app/map/fabric-map.directive.js]: Can be removed.

File: src/main/webapp/app/map/fabric-map.directive.js [hunk 10]
```diff
@@ -187,18 +251,28 @@
...
+                });
                 console.log('vm.selected: ', scope.vm.selected);
             }
 
+            /**
+             * Updates the cameras and players when they are modified in the view.
+             * @param target the camera or player that is updated
+             */
             function onModify(target) {
                 console.log('updated target: ', target);
```
Comment (cadw) [inline src/main/webapp/app/map/fabric-map.directive.js]: Can be removed

File: src/main/webapp/app/map/map-table.directive.js [hunk 11]
```diff
@@ -4,31 +4,49 @@
...
         return directive;
 
+        /**
+         * Links the controller to this directive.
+         * @param scope the scope that the directive is called in
+         * @param element the element that the directive is called in
+         * @param attrs the attributes of the element that the directive is called in
+         */
         function link(scope, element, attrs) {
             console.log('Link from directive map-table-view called');
```
Comment (cadw) [inline src/main/webapp/app/map/map-table.directive.js]: These can be removed.

File: src/main/webapp/app/scripting/scripting.controller.js [hunk 12]
```diff
@@ -17,14 +17,26 @@
      */
     function ScriptingController ($rootScope, $scope, $state, Map, Cue, AlertService) {
         var vm = this;
+        console.log('Scripting controller scope: ', $scope);
```
Comment (cadw) [inline src/main/webapp/app/scripting/scripting.controller.js]: Can be removed.

File: src/main/webapp/app/scripting/scripting.controller.js [hunk 13]
```diff
@@ -17,14 +17,26 @@
...
+        update();
 
-        function getMapEntities() {
-            Map.getDTO({id: 1}, function (result) {
-                console.log('result is: ', result);
-                vm.map = result;
+        $scope.$watch('vm.selected', function (selected) {
+            if(selected.hasOwnProperty('cameraType')) {
+                vm.selectedCamera = selected;
+                console.log('camera selected');
```
Comment (cadw) [inline src/main/webapp/app/scripting/scripting.controller.js]: Can be removed.

File: src/main/webapp/app/scripting/scripting.controller.js [hunk 14]
```diff
@@ -17,14 +17,26 @@
...
+            } else {
+                vm.selectedPlayer = selected;
+                console.log('player selected');
```
Comment (cadw) [inline src/main/webapp/app/scripting/scripting.controller.js]: Can be removed.

File: src/main/webapp/app/scripting/timeline.directive.js [hunk 15]
```diff
@@ -23,45 +34,65 @@
             console.log('Element is: ', element);
```
Comment (cadw) [inline src/main/webapp/app/scripting/timeline.directive.js]: These can be removed.

File: src/main/webapp/app/scripting/timeline.directive.js [hunk 16]
```diff
@@ -23,45 +34,65 @@
...
             console.log('Scope is: ', scope);
 
+            scope.timeline = {};
+            scope.vm.timelineSelected = {};
+            scope.dataset = new vis.DataSet();
+
+            init();
+
             scope.$watch('vm.map', function (newMap) {
                 console.log('Changed map:', newMap);
```
Comment (cadw) [inline src/main/webapp/app/scripting/timeline.directive.js]: Can be removed.

File: src/main/webapp/app/scripting/timeline.directive.js [hunk 17]
```diff
@@ -113,12 +151,41 @@
...
+                        }
+                    }).result.then(function(result) {
+                        var end = result.bar + result.duration;
+
+                        item.content = result.action;
+                        item.start = parseIntAsDate(result.bar);
+                        item.end = parseIntAsDate(end);
+                        item.cue = result;
+
+                        console.log('StartYear is: ', item.start);
```
Comment (cadw) [inline src/main/webapp/app/scripting/timeline.directive.js]: These can be removed.

File: src/main/webapp/app/scripting/scripting-new-dialog.controller.js [hunk 18]
```diff
@@ -7,6 +7,14 @@
 
     ScriptingNewDialogController.$inject = ['$timeout', '$scope', '$uibModalInstance', 'entity'];
 
+    /**
+     * The controller for for the state that adds a new cue.
```
Comment (trombomart) [inline src/main/webapp/app/scripting/scripting-new-dialog.controller.js]: Just a spelling correction: double for

File: src/main/webapp/app/scripting/timeline.directive.js [hunk 19]
```diff
@@ -23,45 +34,65 @@
...
+            scope.$watch('vm.cues', function (newCues) {
+                console.log('Cues changed: ', newCues);
+                createItems(scope.vm.cues);
+            });
+
+            scope.timeline.on('select', function (properties) {
+                console.log('selected timeline items: ', properties);
+            });
+
+            ///////////////////////////////////////////
```
Comment (trombomart) [inline src/main/webapp/app/scripting/timeline.directive.js]: This can be removed?

File: src/main/webapp/app/map/fabric-map.directive.js (hunk 6 above)
Comment (rubenwiersma) [inline src/main/webapp/app/map/fabric-map.directive.js]: This width is not related to the grid

File: src/main/webapp/app/map/fabric-map.directive.js (hunk 7 above)
Comment (rubenwiersma) [inline src/main/webapp/app/map/fabric-map.directive.js]: We've looked into a way to do this, but this resulted in a lot of callbacks that need to be resolved. For now, this is a good placeholder as we might change this to be a shape after all

File: src/main/webapp/app/scripting/timeline.directive.js (hunk 19 above)
Comment (hungnguyen94) [inline src/main/webapp/app/scripting/timeline.directive.js]: It's nice to have some separation between function declarations and other stuff, in my opinion. The style guide also does this.

File: src/main/webapp/app/map/fabric-map.directive.js (hunk 4 above)
Comment (rubenwiersma) [inline src/main/webapp/app/map/fabric-map.directive.js]: Red means that is was deleted

File: src/main/webapp/app/map/fabric-map.directive.js [hunk 20]
```diff
@@ -68,48 +73,71 @@
...
-                console.log('parent is ', parent);
+
+                draw(scope.vm.map.cameras, scope.vm.map.players);
             }
 
+            /**
+             * Prepares the cameras and players for drawing and draws them to the map.
+             * @param cameras the cameras to draw
+             * @param players the players to draw
+             */
```
Comment (ChrisBerg) [inline src/main/webapp/app/map/fabric-map.directive.js]: Nice, comments

**Comment by ChrisBerg** [comment]
Nice work :+1: Merging it into dev.
